test/mongoose_rabbit_worker_SUITE.erl

```diff
@@ -0,0 +1,205 @@
+    Worker = proplists:get_value(worker_pid, Config),
+    Exception = exception_fun(),
+    ConnectionAndChannel0 = get_worker_conn_and_chann(Worker),
+
+    %% when amqp_publish fails
+    gen_server:cast(Worker, {amqp_publish, {Exception, [ok]}, ok}),
+    %% then
+    ConnectionAndChannel1 = get_worker_conn_and_chann(Worker),
+    ?assertNotMatch(ConnectionAndChannel0, ConnectionAndChannel1),
+
+    %% when amqp_call fails
+    gen_server:call(Worker, {amqp_call, {Exception, [ok]}}),
+
+    %% then
+    ConnectionAndChannel2 = get_worker_conn_and_chann(Worker),
+    ?assertNotMatch(ConnectionAndChannel1, ConnectionAndChannel2).
+
+
+worker_processes_msgs_when_queue_msg_len_limit_is_not_reached(Config) ->
+    %% given
+    Worker = proplists:get_value(worker_pid, Config),
+    Ref = make_ref(),
+    Lock = lock_fun(),
+    SendBack = send_back_fun(),
+
+    %% when
+    gen_server:cast(Worker, {amqp_publish, {Lock, [Ref]}, ok}),
+    gen_server:cast(Worker, {amqp_publish, {SendBack, [self(), Ref]}, ok}),
+    [gen_server:cast(Worker, {amqp_publish, ok, ok})
+     || _ <- lists:seq(1, ?MAX_QUEUE_LEN-1)],
+
+    %% unlock the worker
+    Worker ! Ref,
+
+    %% then
+    ?assertReceivedMatch(Ref, 100).
+
+worker_drops_msgs_when_queue_msg_len_limit_is_reached(Config) ->
+    %% given
+    Worker = proplists:get_value(worker_pid, Config),
+    Ref = make_ref(),
+    Lock = lock_fun(),
+    SendBack = send_back_fun(),
+
+    %% when
+    gen_server:cast(Worker, {amqp_publish, {Lock, [Ref]}, ok}),
+    gen_server:cast(Worker, {amqp_publish, {SendBack, [self(), Ref]}, ok}),
+    [gen_server:cast(Worker, {amqp_publish, ok, ok})
+     || _ <- lists:seq(1, ?MAX_QUEUE_LEN+1)],
+
+    %% unlock the worker
+    Worker ! Ref,
+
+    %% then
+    ?assertError({assertReceivedMatch_failed, _},
+                 ?assertReceivedMatch(Ref, 100)).
+
+%%--------------------------------------------------------------------
+%% Helpers
+%%--------------------------------------------------------------------
+
+mock_amqp() ->
+    [meck:new(M, [no_link, passthrough]) || M <- ?AMQP_MOCK_MODULES],
+    meck:expect(amqp_connection, start, fun(_) -> {ok, random_pid()} end),
+    meck:expect(amqp_connection, open_channel, fun(_) -> {ok, random_pid()} end),
+    meck:expect(amqp_connection, close, fun(_) -> ok end),
+    meck:expect(amqp_channel, close, fun(_) -> ok end),
+    meck:expect(amqp_channel, call,
+                fun(_, {F, A}) when is_function(F) -> apply(F, A);
+                   (_, _) -> ok
+                end),
+    meck:expect(amqp_channel, call,
+                fun(_, {F, A}, _) when is_function(F) -> apply(F, A);
+                   (_, _, _) -> ok
+                end).
+
+mock_metrics() ->
+    meck:new(mongoose_metrics, [no_link, passthrough]),
+    meck:expect(mongoose_metrics, update, fun(_, _, _) -> ok end).
+
+unload_amqp() ->
+    [meck:unload(M) || M <- ?AMQP_MOCK_MODULES].
+
+random_pid() ->
+    spawn(fun() -> ok end).
+
+lock_fun() ->
```

Comment:
A bit of an overkill. Absolutely no gain from extracting it into separate function.

Reply:
AFAIR I've put this funs into separate functions for the sake of tracing (I had some troubles when I was developing this suite). I can trace functions from the suite that wrap funs. Couldn't it be helpful? What do you reckon?

Reply:
OK, I accept such explanation for this and other functions. :) I think it would be nice to have a comment here like "BIFs extracted for tracing"